L63-core-call-credentials-debug-string.md

```diff
@@ -0,0 +1,121 @@
+Title
+----
+* Author(s): mhaidry
+* Approver: a11r
+* Status: Draft
+* Implemented in: cpp
+* Last updated: [Date]
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+Implement a DebugString method for the CallCredentials class that returns a
+string sharing details about credentials based on the type of credentials being
+used on the RPC.
+
+## Background
+
+This feature satisfies requests from internal google users who would wish to
+get a debug string from the CallCredentials class
+
+### Related Proposals: 
+None
+
+## Proposal
+Add an abstract DebugString method to the CallCredentials class and plumb the
+implementation of the method to all credential types. The output string depends
+on the kind of credential that is being parsed. The virtual method in the base
+class will have a default implementation so that it does not break any other
+custom credentials.
+
+## Rationale
+
+This provides a way to the user to learn more about the credentials they have
+set for a RPC. It may help with debugging any credentials related issues while
+sending the RPC. The design will also try to make sure that most of the
+computation and memory allocations is done during the DebugString method so that
+there is no default performance hits on the RPC.
+
+## Implementation
+
+### CallCredentials and SecureCallCredentials
+CallCredentials is the primary abstract class. Here we will add a virtual
+DebugString() method that returns an empty string by default.
+
+SecureCallCredentials inherits CallCredentials and will have an implementation
+that relies on its underlying grpc_call_credentials structure to get the
+debug string contents
+
+![CallCredentials class hierarchy](L63_graphics/call_creds_class_hierarchy.png)
+
+### C Call credentials and its derived classes
+The base C struct grpc_call_credentials will provide a helper debug_string().
+This will again be a virtual class that returns an empty string by default.
+We will discuss how every derived grpc_call_credentials struct will implement
+the debug_string method.  For now we will exclude grpc_plugin_credentials and
+grpc_composite_call_credentials. They will be discussed in separate sections
+
+![C CallCredentials hierarchy](L63_graphics/c_call_creds_hierarchy.png)
```

Comment:
Broken link to image?

Reply:
That is strange. I can view the image in my branch
https://github.com/mhaidrygoog/proposal/blob/master/L63-core-call-credentials-debug-string.md

Reply:
I see it when I click "view file" on the file header.

Reply:
Hunh. 🤷‍♂️